cmd/root.go

```
@@ -31,28 +33,59 @@ import (
 )
 
 var (
-	cfgFile                string
-	cpuprofile, memprofile string
+	cfgFile                        string
+	cpuprofile, memprofile         string
+	cpuprofileFile, memprofileFile *os.File
 )
 
 // rootCmd represents the base command when called without any subcommands
 var rootCmd = &cobra.Command{
-	Use:     "hubble",
-	Short:   "CLI",
-	Long:    `Hubble is a utility to observe and inspect recent Cilium routed traffic in a cluster.`,
-	Version: pkg.Version,
+	Use:          "hubble",
+	Short:        "CLI",
+	Long:         `Hubble is a utility to observe and inspect recent Cilium routed traffic in a cluster.`,
+	SilenceUsage: true,
+	Version:      pkg.Version,
+	PersistentPreRunE: func(_ *cobra.Command, _ []string) error {
+		var err error
+		if cpuprofile != "" {
+			cpuprofileFile, err = os.Create(cpuprofile)
+			if err != nil {
+				return fmt.Errorf("failed to create cpu profile: %v", err)
+			}
+			pprof.StartCPUProfile(cpuprofileFile)
+		}
+		if memprofile != "" {
+			memprofileFile, err = os.Create(memprofile)
+			if err != nil {
+				return fmt.Errorf("failed to create memory profile: %v", err)
+			}
+		}
+		return nil
+	},
+	PersistentPostRunE: func(_ *cobra.Command, _ []string) error {
+		if cpuprofileFile != nil {
+			pprof.StopCPUProfile()
+			cpuprofileFile.Close()
+		}
+		if memprofileFile != nil {
+			runtime.GC() // get up-to-date statistics
+			if err := pprof.WriteHeapProfile(memprofileFile); err != nil {
+				return fmt.Errorf("failed to write memory profile: %v", err)
+			}
+			memprofileFile.Close()
+		}
+		return nil
+	},
 }
 
 // Execute adds all child commands to the root command sets flags appropriately.
 // This is called by main.main(). It only needs to happen once to the rootCmd.
-func Execute() {
-	fin := maybeProfile()
-	defer fin() // make sure update memory profile is written in the end
-
-	if err := rootCmd.Execute(); err != nil {
+func Execute() error {
+	err := rootCmd.Execute()
+	if err != nil {
 		fmt.Println(err)
```

it's better to return the error and let the caller choose print or not